src/qt3utils/datagenerators/plescanner.py

```diff
@@ -0,0 +1,499 @@
+import logging
+import time
+import tkinter as tk
+from typing import Type
+
+import numpy as np
+
+from qt3utils.applications.controllers.nidaqedgecounter import QT3ScanNIDAQEdgeCounterController, QT3PleNIDAQEdgeCounterController
+from qt3utils.applications.controllers.wavemeter_controller import WavemeterController
+from qt3utils.applications.controllers.lockin_controller import Lockin
+from qt3utils.nidaq.customcontrollers import WavelengthControlBase, VControl
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+class PleScanner:
+    '''
+    Base class with basic scanning PLE functionality. 
+
+    Current PLE implementation consists of sequential "fast" scans by sweeping
+    the laser frequency via the `wavelength_controller` agrument while simultaneously
+    reading in data via the DAQ (or other generic input source) via the initialization
+    argument `readers`.
+        Currently only a basic DAQ reader is implemented. Additional readers for 
+        simultaneous processing will require additional development at this time.
+    '''
+
+    def __init__(self, 
+                 readers: dict, 
+                 wavelength_controller: WavelengthControlBase,
+                 auxiliary_controllers: dict = {}) -> None:
+        '''
+        Args:
+            readers                 :   Dictionary of data generators, class can be arbitrary
+                                        but should be used to collect data on each scan.
+                                        Later implementations may split this into a principal
+                                        reader (to control timing) and sub-readers for
+                                        additional data collection.
+            wavelength_controller   :   Wavelength controller class inheriting from
+                                        qt3utils.nidaq.customcontrollers.WavelengthControlBase
+            auxiliary_controllers   :   A dictionary of additional controllers of arbitrary
+                                        class. Use this to pass additional controllers for
+                                        intermidiate steps such as pulse sequencing.
+                                        A basic form of unconditional repump is implemented here.
+        
```

Comment:
This is great and was a comment I was going to make throughout, we should be adding docstrings for every function in the future.. Maybe we can bring this up during the group retreat. With GitHub copilot, generating docstrings is usually pretty easy.